odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1471 @@
+                            'not match'.format(func1.range, func2.range))
+
+        Functional.__init__(self, domain=func1.domain,
+                            linear=(func1.is_linear and func2.is_linear),
+                            smooth=(func1.is_smooth and func2.is_smooth),
+                            concave=(func1.is_concave and func2.is_concave),
+                            convex=(func1.is_convex and func2.is_convex),
+                            grad_lipschitz=(func1.grad_lipschitz +
+                                            func2.grad_lipschitz))
+
+        OperatorSum.__init__(self, func1, func2, tmp_ran=None,
+                             tmp_dom=tmp_dom)
+
+    @property
+    def gradient(self):
+        """Gradient operator of functional sum."""
+        return self.left.gradient + self.right.gradient
+
+    def derivative(self, point):
+        """Returns the derivative operator in the given point.
+
+        This function returns the linear operator
+
+            ``x --> <x, grad_f(point)>``,
+
+        where ``grad_f(point)`` is the gradient of the functional in the point
+        ``point``.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        return self._op1.derivative(point) + self._op2.derivative(point)
+
+
+class FunctionalScalarSum(Functional, OperatorSum):
+    """Expression type for the sum of a functional and a scalar.
+
+    ``FunctionalScalarSum(func, scalar) <==> (x --> func(x) + scalar)``
+    """
+    def __init__(self, func, scalar, tmp_dom=None):
+        """Initialize a FunctionalScalarSum instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            Functional to which the scalar is added.
+        scalar : `element` in the `field` of the ``domain``
+            The scalar to be added to the functional. The `field` of the
+            ``domain`` is the range of the functional.
+        tmp_dom : `LinearSpaceVector`, optional
+            ...
+        """
+        # TODO: what is tmp_dom used for?
+        if not isinstance(func, Functional):
+            raise TypeError('functional {!r} is no a Functional instance'
+                            ''.format(func))
+        if scalar not in func.range:
+            raise TypeError('the scalar {} is not in the range of the'
+                            'functional {!r}'.format(scalar, func))
+
+        Functional.__init__(self, domain=func.domain, linear=func.is_linear,
+                            smooth=func.is_smooth, concave=func.is_concave,
+                            convex=func.is_convex,
+                            grad_lipschitz=func.grad_lipschitz)
+
+        OperatorSum.__init__(self, left=func,
+                             right=ConstantOperator(vector=scalar,
```

Comment:
Is there any performance gains/losses depending on weather the functional is `left`or `right` in `OperatorSum`?